### DISCUSSION TITLE
RPC tutorial questions regarding the queue and channel

### DISCUSSION THREAD

**Discussion by wigging** [rabbitmq/rabbitmq-server#12193] [Questions] [answered]
I went through the [RPC tutorial](https://www.rabbitmq.com/tutorials/tutorial-six-python) on the RabbitMQ website and have some questions.

Why does the client code (see below) not define the queue name? For example, instead of using `queue_declare(queue='', exclusive=True)`, it seems like you could just do something like `queue_declare(queue='client_queue', exclusive=True)` where the name of the queue is explicitly defined.
```python
result = channel.queue_declare(queue='', exclusive=True)
callback_queue = result.method.queue
```

In the server portion of the code (see below), which channel `ch` is being used in the callback function? Is it the same channel such as `channel = connection.channel()` that is defined earlier in the code?
```python
def on_request(ch, method, props, body):
    ...
```

**Comment by michaelklishin** [comment, marked as answer]
@wigging you can use an client-provided queue name but since that is a response queue, that won't work very well once you have N clients. Therefore [server-named queues](https://www.rabbitmq.com/docs/queues#server-named-queues) are used for response (request-reply client-owned) in the tutorial.

This is different from request-reply server queues which can and usually are explicitly named and "well-known" in the overall system. N competing consumers (server instances) on such queues pose no problem and create no confusion.

The channel used in Pika's callbacks is the channel the delivery arrived on, which is the channel you have registered the consumer on.

**Reply by michaelklishin** [reply]
I'd also add that https://github.com/rabbitmq/rabbitmq-server/discussions/12181#discussioncomment-10508001 is relevant to the request-reply messaging pattern, and provides another sound justification for the use of server-named temporary queues for reply collection.

**Reply by wigging** [reply]
Ok, so when I have multiple clients they can't use the same response queue; therefore, the tutorial uses a server-named queue. But why can't multiple clients use the same queue? It seems like using the same response queue would be more efficient than creating a new one for each client.

**Reply by michaelklishin** [reply]
For the same reason why not having shared state is often necessary or a very good idea. How would they determine which delivery is theirs? Yes, using `correlation_id` and other metadata fields, but they they would have to requeue most of deliveries.

In AMQP 0-9-1, the idea that one client often means one queue, and that eliminates the need for a "query language" and so on, is everywhere.

The tutorials do not use this feature because their goal is to demonstrate the concept but for 11 years now, there [exists a way to not declare any response queues](https://www.rabbitmq.com/docs/direct-reply-to) specifically for this pattern. It has its downside of not being a "real queue" but for transient responses that's usually fine, or you can do what tutorial 6 does, or even try what you had in mind (now being aware of the risks and complexities of such shared client state).

**Reply by wigging** [reply]
Thank you for the explanation.
